### ISSUE TITLE
Firmware issues with new RT3s models

### ISSUE THREAD

**Issue by silseva** [OpenRTX/OpenRTX#198]
DB2OE and ON4MOD reported the following issues affecting OpenRTX when flashed on new Retevis RT3s radios:

1. settings not persistent, except the first time the radio is powered up after a firmware update
2. missing TX in FM mode
3. firmware freezing when switching to M17 mode 

The first problem may be related to the flash requiring a full sector erase before each write.

**Comment by silseva** [comment]
The discussion here provides some hits, as the symptoms are similar: https://www.opengd77.com/viewtopic.php?t=2667
The MCUs installed in the new RT3s may be APM32F405: https://global.geehy.com/apm32?id=49

**Comment by silseva** [comment]
Test firmware to try tackle point 3: hw version field of info page contains the DBGMCU_IDCODE, interrupt level of audio stream drivers raised from 10 to 7.

[openrtx_mduv3x0_test.zip](https://github.com/OpenRTX/OpenRTX/files/12841594/openrtx_mduv3x0_test.zip)

**Comment by UnImAtrIXzEr0** [comment]
I now believe there is a problem with radio_tool under Linux. As soon as I write with the manufacturer's flash tool under Windows 10, the device with M17 does not freeze. FM transmission also works perfectly. That's why we suspect that the radio_tool and the manufacturer's software flash slightly differently.

So I don't think errors 2 and 3 are a firmware problem.

But point 1 remains, saving doesn't work properly yet.

**Comment by silseva** [comment]
This is very interesting. Just to have another test, can you flash the latest nightly build with both the OEM firmware update tool and `radio_tool`, please?
We'll also organize to buy a new UV380/RT3s in order to have a sample of the latest hardware.

**Comment by UnImAtrIXzEr0** [comment]
Ok, I installed the nightly openrtx_mduv3x0_wrap from october 8th with radio_tool. 
Now the radio works just before flashed with OEM firmware update tool. The strange thing is that after flashing the settings are still there, such as callsign and frequency.
Good up to this point.

**Comment by silseva** [comment]
After some investigations with Morgan ON4MOD we found out that there is something weird going on when writing to the MCU flash memory - were we're storing the settings and persistence now - and it also seems that the new MD-UV380/RT3s have an APM32F4. From the tests we run, it seems that only the first byte gets written to the flash and then the memory goes into an unknown state were it does not write anything else. The only possible solution to this issue, right now, is to do a full sector erase before each write, which is something that reduces the lifespan of the flash memory.

Given that soon we'll have codeplug support on the external flash and that we'll also move the persistence there, I'll prefer to invest time on that rather than trying to find a solution for the non-persistent settings issue, also because I don't have a recent RT3s.

**Comment by iz4tow** [comment]
so? can you provide the istructions? Thanks

**Comment by kosma** [comment]
I have tracked down this issue to the FLASH programming registers being very different on APM32F405. Resolving this will require writing a separate FLASH driver for those MCUs, and either detecting the MCU model during runtime, or having separate builds for different MCUs. Compare [RM0090](https://www.st.com/resource/en/reference_manual/rm0090-stm32f405415-stm32f407417-stm32f427437-and-stm32f429439-advanced-armbased-32bit-mcus-stmicroelectronics.pdf) pages 99-113 and [APM32F403xB UM](https://www.st.com/resource/en/reference_manual/rm0090-stm32f405415-stm32f407417-stm32f427437-and-stm32f429439-advanced-armbased-32bit-mcus-stmicroelectronics.pdf) pages 21-24. Also pinging issue #325.

**Comment by silseva** [comment]
Thanks for the research!
Actually I prefer to have a solution based on runtime detection of the MCU type, hoping that the content of the CPUID register is different between the STM32 and APM32 devices...
The thing is that we're going towards a solution using the external flash memory to store the device settings and persistent state, so MCU flash will be abandoned at some point. However, given that I cannot predict how much time this migration work is going to take, having the current driver fixed is also good. Do you want to take care of it?

**Comment by kosma** [comment]
> Do you want to take care of it?

Absolutely. My friends are getting me a 1701 with a clone CPU to debug the issue, we're starting work tomorrow. :)

**Comment by lemielek** [comment]
@silseva I have fresh DM-1701, S/N: 24DM170112717, FW VER: V002.006, CP VER: V01.05 and I am wondering if copy of security registers I made from my RT3S for OpenRTX-external-docs is correct or should be done again after this MCU driver correction? Also same for this new Baofeng.

**Comment by chrisdebian** [comment]
This has been open for a while without further discussion, is this still reproducible on a current build? If so, it would be useful to know the flash chip fitted in the affected radios (readable via the SPI flash ID registers or by opening the radio). If newer RT3s units use a flash that requires a full sector erase before write, the NVM driver would need updating to handle that; a different erase granularity would explain the settings persistence issue specifically. The FM TX and M17 freeze problems may be separate or may also be hardware-revision related. Any additional detail from DB2OE or ON4MOD would help narrow this down.

**Comment by silseva** [comment]
Points 2) and 3) are solved, point 1) will be solved with release v0.5.0 by moving all the settings storage to the external SPI flash.

**Comment by sdresen** [comment]
New to M17 and trying to flash OpenRTX onto a brand new Retevis RT3S.  After flash, still running stock firmware.  Claude helped me write the following info should it be helpful.  Assuming there will be a fix eventually for this but reporting should it be helpful.

Radio: Retevis RT3S
OpenRTX version: v0.4.4, openrtx_mduv3x0_v0.4.4.bin
Tools: radio_tool (built from source, latest main), dfu-util 0.11
OS: Linux

Issue: Flashing OpenRTX via radio_tool reports success, but the radio still boots the original stock firmware afterward — no OpenRTX UI, no change in behavior.

Details:
dfu-util -l shows three DFU alt-settings on this unit:

alt=0, @Internal Flash, 0x08000000, 03*016Kg (only ~48KB exposed)
alt=1, @1G Flash Memory, 0x10000000, 512*256Kg (~128MB)
alt=2, @SPI Flash Memory, 0x00000000, 16*064Kg (~1MB)

radio_tool -d 0 -f -i openrtx_mduv3x0_v0.4.4.bin erased and wrote sectors starting at 0x0800c000 through ~0x08045bb0 (matches the 231KB image size exactly, transfer completes cleanly with "Done!").

Despite the successful-looking write to internal flash, the radio boots stock firmware unchanged afterward. This suggests my unit may be one of the "newer RT3S" hardware revisions mentioned in #198, where the application firmware may actually execute from the external SPI flash region rather than internal MCU flash, making radio_tool's internal-flash write a no-op for what actually boots.

I took full backups of all three flash regions before flashing, so the radio is not bricked — just still running stock firmware.

Question: Is this a known/supported hardware variant? If the app should be targeting the SPI flash region instead, is there a known correct offset/format for that, or is this hardware revision not yet supported by OpenRTX/radio_tool?

Full debug output attached.


[rt3s_openrtx_debug.txt](https://github.com/user-attachments/files/30917219/rt3s_openrtx_debug.txt)

**Comment by sdresen** [comment]
Following up, I used [https://dmr.tools](https://dmr.tools) and was able to successfully flash OpenRTX without issue.
